crates/fuel-core/src/coins_query.rs

```diff
@@ -221,6 +274,168 @@ pub async fn random_improve(
+    const TOTAL_AMOUNT_ADJUSTMENT_FACTOR: u64 = 2;
+    if total == 0 && max == 0 {
+        return Ok(vec![]);
+    }
+
+    let adjusted_total = total.saturating_mul(TOTAL_AMOUNT_ADJUSTMENT_FACTOR);
+
+    let big_coins_stream = futures::stream::iter(big_coins_iter).yield_each(batch_size);
+    let dust_coins_stream = futures::stream::iter(dust_coins_iter).yield_each(batch_size);
+
+    let (selected_big_coins_total, selected_big_coins) =
+        big_coins(big_coins_stream, adjusted_total, max, excluded_ids).await?;
+
+    if selected_big_coins_total < total {
+        return Ok(vec![]);
+    }
+    let Some(last_selected_big_coin) = selected_big_coins.last() else {
+        // Should never happen.
+        return Ok(vec![]);
+    };
+
+    let selected_big_coins_len = selected_big_coins.len();
+    let number_of_big_coins: u16 = selected_big_coins_len.try_into().map_err(|_| {
+        CoinsQueryError::TooManyCoinsSelected {
+            required: selected_big_coins_len,
+            max: u16::MAX,
+        }
+    })?;
+
+    let max_dust_count = max_dust_count(max, number_of_big_coins);
+    let (dust_coins_total, selected_dust_coins) = dust_coins(
+        dust_coins_stream,
+        last_selected_big_coin,
+        max_dust_count,
+        excluded_ids,
+    )
+    .await?;
+    let retained_big_coins_iter =
+        skip_big_coins_up_to_amount(selected_big_coins, dust_coins_total);
+
+    Ok((retained_big_coins_iter.chain(selected_dust_coins)).collect())
+}
+
+async fn big_coins(
+    big_coins_stream: impl Stream<Item = StorageResult<CoinsToSpendIndexEntry>> + Unpin,
+    total: u64,
+    max: u16,
+    excluded_ids: &ExcludedCoinIds<'_>,
+) -> Result<(u64, Vec<CoinsToSpendIndexEntry>), CoinsQueryError> {
+    select_coins_until(big_coins_stream, max, excluded_ids, |_, total_so_far| {
+        total_so_far >= total
+    })
+    .await
+}
+
+async fn dust_coins(
+    dust_coins_stream: impl Stream<Item = StorageResult<CoinsToSpendIndexEntry>> + Unpin,
+    last_big_coin: &CoinsToSpendIndexEntry,
+    max_dust_count: u16,
+    excluded_ids: &ExcludedCoinIds<'_>,
+) -> Result<(u64, Vec<CoinsToSpendIndexEntry>), CoinsQueryError> {
+    select_coins_until(
+        dust_coins_stream,
+        max_dust_count,
+        excluded_ids,
+        |coin, _| coin == last_big_coin,
+    )
+    .await
+}
+
+async fn select_coins_until<F>(
+    mut coins_stream: impl Stream<Item = StorageResult<CoinsToSpendIndexEntry>> + Unpin,
+    max: u16,
+    excluded_ids: &ExcludedCoinIds<'_>,
+    predicate: F,
+) -> Result<(u64, Vec<CoinsToSpendIndexEntry>), CoinsQueryError>
+where
+    F: Fn(&CoinsToSpendIndexEntry, u64) -> bool,
+{
+    let mut coins_total_value: u64 = 0;
+    let mut count = 0;
+    let mut coins = Vec::with_capacity(max as usize);
+    while let Some(coin) = coins_stream.next().await {
+        let coin = coin?;
+        if !is_excluded(&coin, excluded_ids)? {
+            if count >= max || predicate(&coin, coins_total_value) {
```

Comment:
```suggestion
            if coins.len() >= max || predicate(&coin, coins_total_value) {
```

I think we can remove `count`=)

Reply:
Yes, well spotted. Removed in 0602da1a562278603c47443b35cb5d09101f74fd